### PR TITLE
Using `std` maps in datagen

### PR DESCRIPTION
**Pull request by robertbastian**
Currently we're using `LiteMap` with no clear benefit but several drawbacks:
* insertion is less performant than `HashMap` or `BTreeMap`
* The API is less powerful
* Requires `Ord` where `Hash` would do

`HelloWorldDataProvider` can just use a static slice and doesn't have to allocate at all.

### CODE REVIEW

**Review by Manishearth — DISMISSED** [review]

File: provider/testdata/data/baked/datetime/datelengths_v1.rs [hunk 1]
```diff
@@ -1,48 +1,18 @@
...
-    ("ar-u-ca-gregory", AR_U_CA_GREGORY_AR_EG_U_CA_GREGORY),
-    (
-        "ar-u-ca-indian",
-        AR_U_CA_BUDDHIST_AR_U_CA_COPTIC_AR_U_CA_ETHIOPIC,
-    ),
-    (
-        "ar-u-ca-japanese",
-        AR_U_CA_BUDDHIST_AR_U_CA_COPTIC_AR_U_CA_ETHIOPIC,
-    ),
+    ("ar-EG-u-ca-buddhist", AR_EG_U_CA_BUDDHIST_AR_EG_U_CA_COPTIC),
```
Comment (sffc) [inline provider/testdata/data/baked/datetime/datelengths_v1.rs:4]: Question/Issue: Why did the generated data files change

File: provider/core/src/resource.rs [hunk 2]
```diff
@@ -118,6 +118,19 @@ pub struct ResourceKey {
     hash: ResourceKeyHash,
 }
 
+// I don't think this lint makes sense, see https://github.com/rust-lang/rust-clippy/issues/2627.
+// It's not possible to violate the hash invariant (k1 == k2 -> hash(k1) == hash(k2)) without
+// using non-determinism or unsafe code when PartialEq is derived.
+#[allow(clippy::derive_hash_xor_eq)]
+
+// Custom implementation to avoid hashing the path.
+impl core::hash::Hash for ResourceKey {
```
Comment (sffc) [inline provider/core/src/resource.rs:127]: Issue: I feel that `ResourceKeyHash` should be used instead of `ResourceKey` when a hash is needed, which is why there isn't already an `impl Hash for ResourceKey`.

File: provider/datagen/src/transform/cldr/plurals/mod.rs [hunk 3]
```diff
@@ -74,7 +74,7 @@ impl<M: ResourceMarker<Yokeable = PluralRulesV1<'static>>> IterableResourceProvi
         Ok(self
             .get_rules_for(M::KEY)?
             .0
-            .iter_keys()
+            .keys()
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/plurals/mod.rs:77]: Issue, here and elsewhere: one of the motivations for migrating from HashMap to LiteMap (#1341) was that it gives us stable ordering when calling `.iter()`, which is a desirable property. When generating data, you _really_ don't want to introduce entropy based on what state a hash map is in.

**Review by sffc — CHANGES_REQUESTED** [review]
Not sure I like this. I migrated these _from HashMap to LiteMap_ in #1341, and the reasons for doing it then haven't really changed.

**Comment by Manishearth** [comment]
I think the reason was in part to get rid of Ord requirements

File: provider/datagen/src/transform/cldr/plurals/mod.rs (hunk 3 above)
Comment (robertbastian) [reply provider/datagen/src/transform/cldr/plurals/mod.rs:77]: The output of `supported_options` is `par_iter`ed, so there's no point whatsoever in this being stable.

There was only one case where generating code expected a serde map to be sorted and I used a `BTreeMap` for that.

File: provider/core/src/resource.rs (hunk 2 above)
Comment (robertbastian) [reply provider/core/src/resource.rs:127]: But `ResourceKeyHash` doesn't have the path, which I need for databake. This is like saying you should store `f(x)` is a hash set instead of `x`.

File: provider/testdata/data/baked/datetime/datelengths_v1.rs (hunk 1 above)
Comment (robertbastian) [reply provider/testdata/data/baked/datetime/datelengths_v1.rs:4]: The names of some statics changed because before I sorted on `ResourceOption` and now I'm sorting on their string representation (and the names are concatenations of the first few options).

**Comment by robertbastian** [comment]
One of the reasons you mention in #1341 was that we can add methods to `LiteMap`, but this PR shows that there's nothing that `HashMap` can't do. In fact `HashMap` (and `BTreeMap`) have more powerful APIs, like the `entry` methods that allow update-or-insert semantics. 

The only valid point in #1341 imho is keeping `LiteMap` from bitrotting, but I don't think it's strong enough to justify jumping through hoops to make it work, because `Ord` is going away.

File: provider/core/src/resource.rs (hunk 2 above)
Comment (sffc) [reply provider/core/src/resource.rs:127]: I cannot approve this PR given that my PR https://github.com/unicode-org/icu4x/pull/2115 conflicts with the `Hash` impl. We can't merge both.

Comment (robertbastian) [reply provider/core/src/resource.rs:127]: What's the conflict exactly?

Comment (sffc) [reply provider/core/src/resource.rs:127]: We need to decide on what to do with the ResourceKeyMetadata.

1. Does it affect PartialEq? (then it needs to go into Hash)
2. Is it ignored for PartialEq? (then match statements get less efficient)
3. Does it go into the path string itself? (then it gets into Hash and PartialEq automatically)

I don't want to lock ourselves into having a Hash impl if the impl would be inefficient.

tl;dr, let's reach consensus on #2115 and then we can move on.

File: provider/testdata/data/baked/datetime/datelengths_v1.rs (hunk 1 above)
Comment (sffc) [reply provider/testdata/data/baked/datetime/datelengths_v1.rs:4]: OK.

Previously you had `AR_U_CA_BUDDHIST_AR_U_CA_COPTIC_AR_U_CA_ETHIOPIC`

And now you have `AR_EG_U_CA_BUDDHIST_AR_EG_U_CA_COPTIC`

I was wondering how that went from 3 locales to 2 locales but it looks like it is because of

```rust
                    // Cap identifier length at around 35
                    if a.len() < 35 {
                        a.push('_');
                        a.push_str(&b);
                    }
```

And the new locales are longer, so they stop at 2 instead of 3.

File: provider/datagen/src/transform/cldr/plurals/mod.rs (hunk 3 above)
Comment (sffc) [reply provider/datagen/src/transform/cldr/plurals/mod.rs:77]: ok; I'm a little worried that there are `.iter()` calls elsewhere in datagen that assumed that things were sorted. You found one of them, presumably because there was a test/diff that caught it, but how confident are we that there aren't more? To me this is a problem that should be handled by having separate method names, like `.iter_sorted()` that LiteMap and BTreeMap implement but not HashMap. That said, this is only a theoretical concern short of double-checking all of the code in datagen.

**Review by sffc — COMMENTED** [review]
Once we decide on how to do metadata in ResourceKey, I'm okay merging this.

File: provider/core/src/resource.rs (hunk 2 above)
Comment (robertbastian) [reply provider/core/src/resource.rs:127]: This is independent of #2115. Hashing on hash (aka path) only is valid, even if PartialEq compares more fields. Hash is free to produce collisions as long as it produces equal hashes for PartialEq values. Hash collisions don't affect correctness, they're a performance problem, but that doesn't matter here because a) it's only used in datagen and b) we don't have keys that differ only by metadata anyway.

File: provider/datagen/src/transform/cldr/plurals/mod.rs (hunk 3 above)
Comment (robertbastian) [reply provider/datagen/src/transform/cldr/plurals/mod.rs:77]: I actually double checked most of the code, but not all (newtype structs are annoying because you can search for field names).

File: provider/core/src/resource.rs (hunk 2 above)
Comment (sffc) [reply provider/core/src/resource.rs:127]: ok.

**Review by sffc — APPROVED** [review]
.
